### ISSUE TITLE
Option to defer validation to class instantiation

### ISSUE THREAD

**Issue by ashwin153** [mkorpela/overrides#63]
In my project, I have a class hierarchy that resembles the following.

```python
from __future__ import annotations

import overrides

class Parent(overrides.EnforceOverrides):
  def child(self) -> Child:
    raise NotImplementedError

class Child(Parent):
  @overrides.overrides
  def child(self) -> Child:
    return self
```

Since overrides are validated on class creation and not instantiation (which is a good thing and should remain the default), the type `Child` is not yet defined when the type annotations of `Child.children` are evaluated so an error is raised. It would be cool if there was a way to opt-into deferred validated (e.g., `@overrides.overrides(defer=True)` to defer the validation at the cost of a small performance hit.

**Comment by ashwin153** [comment]
What do you think? @mkorpela

**Comment by mkorpela** [comment]
an option to validate on method execution time?

**Comment by ashwin153** [comment]
Yeah exactly.

**Comment by mkorpela** [comment]
Sounds great :) Name could be something eöse than defer though

**Comment by ashwin153** [comment]
You pick the name, I'll put up an implementation.

**Comment by mkorpela** [comment]
After long process of thinking.. `runtime_validate=True`

**Comment by ashwin153** [comment]
Sounds good. I think https://github.com/mkorpela/overrides/issues/56 one needs to happen first.

**Comment by ashwin153** [comment]
Will do this after you close out that one.

**Comment by mkorpela** [comment]
That will be tomorrow. It is 12 midnight here now.

**Comment by ashwin153** [comment]
Sorry, totally forgot about time zones. There's absolutely no rush. Have a great weekend! Had a blast hacking on this with you the past few days!

**Comment by mkorpela** [comment]
After thinking a bit more, I think you are not thinking of "runtime" validation, but class instantiation time validation.
Maybe `validate_on_class_instantiation=True` ?

It has been nice to work on this project again. Very good work from your side! I've been focusing on other projects for a while.
The things done so far are limited mostly to use of EnforceOverrides class.
`@overrides` decorator use without it is a lot bigger. So consider what has happened so far as beta testing - we have already 2 people giving feedback about bugs.
Package use is rather big https://pypistats.org/packages/overrides - so when now moving to `@overrides` decorator, have to now also handle known problems.

**Comment by brentyi** [comment]
I also have a lot of forward references in code that's being checked, which is currently breaking. Seems like a common use case!

Perhaps type signature checks could be made warnings and/or opt-in until these kinks are sorted out? Or even opt-in permanently -- it's a very useful feature, but also a big breaking change* that (to my understanding) is redundant for folks who already use static type checkers.

*I've only checked two of them, but due to problems with forward references and #64, I suspect that most of my projects that use `EnforceOverrides` are broken now. Current solutions seem to be either disabling enforcement or locking dependencies to `overrides==3.1.0`.

Here's an extra example on what's breaking in Python 3.6+:
```python
import abc
from typing import TypeVar

from overrides import EnforceOverrides, overrides

T = TypeVar("T", bound="Copyable")


class Copyable(abc.ABC, EnforceOverrides):
    @abc.abstractmethod
    def copy(self: T) -> T:
        """Return copy of self"""


class C(Copyable):
    @overrides
    def copy(self) -> "C":
        # Do something
        return self
```

**Comment by ashwin153** [comment]
To my knowledge, mypy will not complain about the following because it does not understand that `Child.foo` is supposed to override `Parent.foo`. In other words, this library provides additional type safety because it can additional check that an method has a compatible signatures with the method that it claims to override.

```python
class Parent:
    def foo(self, x) -> None:
        pass

class Child(Parent):
    def foo(self, y) -> None:
        pass
```

I think locking dependencies to `3.1.0` is a reasonable short-term solution. I have some free time over the next few days, and I plan to spend it knocking out the remaining bugs.

**Comment by brentyi** [comment]
Really appreciate the quick followup. I believe that case won't because of python/mypy#6709.

These examples, however, do:
```
class Parent:
    def foo(self, x: int) -> None:
        pass

class Child(Parent):
    def foo(self, x: str) -> None:
        pass
```

```
class Parent:
    def foo(self, x) -> None:
        pass

class Child(Parent):
    def foo(self, x, y) -> None:
        pass
```

```
class Parent:
    def foo(self, x) -> None:
        pass

class Child(Parent):
    def foo(self, y) -> str:
        return ""
```

**Comment by ashwin153** [comment]
I honestly assumed that this library was validating overrides and not just checking for existence until I overrode a method with a different signature and it blew up at runtime. It is a feature that I think most users would expect of a library that claims to be a drop-in replacement for `overrides` in Java and other languages. While it is good to know that parameter validation is something that is on mypy's radar, I think that it is a big enough limitation to justify this feature. Besides, more static analysis never hurts.

**Comment by brentyi** [comment]
Totally makes sense; appreciate your work on this. Will be watching for updates but I'll pin to `3.1.0` in the meantime.

**Comment by mkorpela** [comment]
@ashwin153 name could also be `forward_reference=True` .. there is one implementation detail that bite me. `sys._getframe` will behave in very different way when the decorator is changed to have arguments. But once you have some solution to this problem, we could see what we can do. One option would be to have another decorator instead of a decorator param.

**Comment by mkorpela** [comment]
This is done in #69

**Comment by brentyi** [comment]
Would it make sense to auto-detect a value for `check_at_runtime`, or otherwise suppress the error message if it's not passed in?

As a user it feels a bit cumbersome to need to always think about which annotations are forward references and which ones aren't; would complicate simple tasks like changing the order of class definitions.

It also looks like there are some edge cases that can't be worked around?
For example, there's no guarantee that forward references are even imported at runtime. This is a very common pattern for avoiding cyclic imports:
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import Something

def x(value: "Something") -> None:
    pass
```

or, hinted functions can be run before the necessary names are even defined:
```
class A():
    @staticmethod
    def function(x: "B") -> None:
        ...

A.function()

class B():
    ...
```
(this is more contrived, but calling static methods at definition time, eg via `__init_subclass__`, is something I've seen a few times)
